tslearn/early_classification.py

```diff
@@ -0,0 +1,312 @@
+
+    Attributes
+    --------------------
+    silhouette_ : float
+        The silhouette score from the clustering
+
+    classifiers_ : list
+        A list containing all the classifiers trained for the model
+
+    clusters_ : dictionary
+        Contains the times series by clusters
+
+    pyhatyck_ : dictionary
+        Contains the probabilities of being classified as class y_hat given class y and cluster ck
+
+    indice_ck_ : list
+        Contains for each clusters the indexes of the time series in the dataset
+
+    __n_classes_ : int private
+        The number of classes in the series
+
+    __len_X_ : int
+        The length of the time series
+
+    Examples
+    --------------------
+
+    References
+    --------------------
+    [1] A. Dachraoui, A. Bondu & A. Cornuéjols. Early classification of time series as a non myopic sequential decision
+    making problem. 2015 Conference paper
+
+
+    """
+
+    def __init__(
+        self,
+        n_clusters,
+        base_classifier,
+        minimum_time_stamp,
+        lamb,
+        cost_time_parameter,
+    ):
+        super(NonMyopicEarlyClassification, self).__init__()
+        self.base_classifier = base_classifier
+        self.n_clusters = n_clusters
+        self.minimum_time_stamp = minimum_time_stamp
+        self.lamb = lamb
+        self.cost_time_parameter = cost_time_parameter
+
+    def fit(self, X, y):
+        """
+        This function fits classifiers that are currently multilayer perceptrons to a training set of time series and
+        associated classes. A classifier is fit for each time stamp above a minimum time stamp that is an attribute of
+        the class of the model. Then some probabilities are computed using clusters already computed.
+        This function should be divided among 2 or 3 functions.
+
+        Parameters
+        ----------
+        X: Array-like
+            a dataset of time series
+        y: vector
+            the associated classes of the series from X_train
+        """
+        y_classes = np.unique(y)
+        self.labels_ = sorted(set(y_classes))
+        y_classes_indices = [self.labels_.index(yi) for yi in y_classes]
+
+        for idx, current_classe in enumerate(y_classes):
+            y[y == current_classe] = y_classes_indices[idx]
+
+        self.cluster_ = TimeSeriesKMeans(n_clusters=self.n_clusters)
+        self.classifiers_ = {
+            t: clone(self.base_classifier)
+            for t in range(self.minimum_time_stamp, X.shape[1] + 1)
+        }
+        self.pyhatyck_ = {}
+        self.indice_ck_ = []
+        self.__n_classes_ = len(y_classes_indices)
+        self.__len_X_ = X.shape[1]
+        c_k = self.cluster_.fit_predict((to_time_series_dataset(X)))
+        X_and_cluster = np.concatenate((X, c_k[:, np.newaxis]), axis=1)
+        X1, X2, y1, y2 = train_test_split(X_and_cluster, y, test_size=0.5)
```

Comment:
`train_test_split` accepts as many arrays as inputs as you want, so you can pass X, y and c_k in and get all of them (doubled) as outputs, instead of this (rather ugly :) hack.